activation/activation.go

```diff
@@ -591,23 +611,26 @@ func (b *Builder) PublishActivationTx(ctx context.Context) error {
 func (b *Builder) createAtx(ctx context.Context) (*types.ActivationTx, error) {
 	b.log.With().Info("challenge ready")
 
-	pubEpoch := b.challenge.PublishEpoch()
-
 	hash, err := b.challenge.Hash()
 	if err != nil {
 		return nil, fmt.Errorf("getting challenge hash failed: %w", err)
 	}
 
 	// the following method waits for a PoET proof, which should take ~1 epoch
-	expireLayer := (pubEpoch + 2).FirstLayer()
-	atxExpired := b.layerClock.AwaitLayer(expireLayer) // this fires when the target epoch is over
+	pubEpoch := b.challenge.PublishEpoch()
+	// Add a 10% margin to the time it takes to generate a PoST proof
+	postDurationWithMargin := time.Duration(float64(b.lastPostGenDuration.Nanoseconds()) * 1.1)
+	postDurationInLayers := b.layerClock.DurationToLayers(postDurationWithMargin)
+	deadlineLayer := b.challenge.PubLayerID.Add(types.GetLayersPerEpoch()).Sub(postDurationInLayers)
+	deadline := b.layerClock.AwaitLayer(deadlineLayer)
 
-	b.log.With().Info("building NIPost", log.Stringer("pub_epoch", pubEpoch), log.Stringer("expire_layer", expireLayer))
+	b.log.With().Info("building NIPost", log.Stringer("pub_epoch", pubEpoch), log.Stringer("deadline_layer", deadlineLayer), log.Uint32("last PoST duration", postDurationInLayers))
 
-	nipost, err := b.nipostBuilder.BuildNIPost(ctx, hash, atxExpired)
+	nipost, postDuration, err := b.nipostBuilder.BuildNIPost(ctx, hash, deadline)
```

Comment:
Instead of returning postDuration here I'd prefer it being a field of `nipostBuilder` that is then read from to set `b.lastPostGenDuration` below or even replace it.

Reply:
I didn't want to keep this information in the state of two structs (`Builder` and `NIPostBuilder`). It needs to be kept in `Builder` because it is initialized with the duration of the first PoST here: https://github.com/spacemeshos/go-spacemesh/blob/2e230ca4d4081e68d517edc538270a5051d3ffa6/activation/activation.go#L320

Therefore, I decided to just return it from `BuildNIPost()` and keep it in a single place (in the state of `Builder`).

Reply:
This line: https://github.com/spacemeshos/go-spacemesh/blob/2e230ca4d4081e68d517edc538270a5051d3ffa6/activation/activation.go#L316

is expected to be called once during the whole lifetime of the node, when creating / publishing its very first ATX. While the call to `BuildNIPost` happens during every epoch. So unless the node has never published an ATX before, this won't be called and `b.lastPostGenDuration` is 0 and later set to the return value of `BuildNIPost()`

Reply:
Yes, it's executed once to calculate the first PoST. However, I still want to use this information to predict the next PoST duration.